cmd/demo/node_setup.go

```diff
@@ -103,63 +105,125 @@ func (n *node) setup() error {
 	return nil
 }
 
-// setupContracts reads from the config file whether the node should deploy or use
-// existing contract addresses.
-func (n *node) setupContracts() (err error) {
-	var adjAddr, assAddr common.Address
+func getTxContext() (context.Context, context.CancelFunc) {
+	return context.WithTimeout(context.Background(), config.Chain.TxTimeout)
+}
 
-	if config.Chain.Adjudicator == "deploy" {
-		adjAddr, err = deployAdjudicator(n.cb)
-		if err != nil {
-			return
-		}
-		fmt.Println("💭 Adjudicator contract deployed")
-	} else {
-		tmpAdj, err := strToAddress(config.Chain.Adjudicator)
-		if err != nil {
-			return err
-		}
-		adjAddr = ewallet.AsEthAddr(tmpAdj)
+func (n *node) validateAdjudicatorContract(adjAddr common.Address) error {
+	ctx, cancel := getTxContext()
+	defer cancel()
+	return echannel.ValidateAdjudicator(ctx, n.cb, adjAddr)
+}
 
-		ctx, cancel := context.WithTimeout(context.Background(), config.Chain.TxTimeout)
-		defer cancel()
-		if err := echannel.ValidateAdjudicator(ctx, n.cb, adjAddr); err != nil {
-			return errors.WithMessage(err, "validating adjudicator contract")
-		}
+func (n *node) validateAssetHolderContract(assAddr common.Address, adjAddr common.Address) error {
+	ctx, cancel := getTxContext()
+	defer cancel()
+	return echannel.ValidateAssetHolderETH(ctx, n.cb, assAddr, adjAddr)
+}
+
+func stringToAddress(s string) (common.Address, error) {
+	var err error
+	var walletAddr wallet.Address
 
-		fmt.Println("💭 Adjudicator contract validated")
+	if walletAddr, err = strToAddress(s); err != nil {
+		return common.Address{}, err
 	}
 
-	if config.Chain.Assetholder == "deploy" {
-		assAddr, err = deployAsset(n.cb, adjAddr)
-		if err != nil {
-			return
-		}
-		fmt.Println("💭 Asset holder contract deployed")
-	} else {
-		tmpAsset, err := strToAddress(config.Chain.Assetholder)
-		if err != nil {
-			return err
+	return ewallet.AsEthAddr(walletAddr), nil
+}
+
+func (n *node) setupAdjudicatorContract() error {
+	// config options: "[onChainAddress]", "deploy", "[onChainAddress]|deploy"
+
+	tokens := strings.Split(config.Chain.Adjudicator, "|")
+
+	for _, token := range tokens {
+		if token == "deploy" {
+			// deploy contract
+
+			if adjAddr, err := deployAdjudicator(n.cb); err == nil {
+				n.adjAddr = adjAddr
+				n.adjudicatorSetup = true
+				fmt.Printf("🌐 Adjudicator contract deployed at %x\n", adjAddr)
+				return nil
+			}
+
+			fmt.Println("❌ Adjudicator contract deployment failed")
+		} else {
+			// validate contract
+
+			if adjAddr, err := stringToAddress(token); err == nil {
+				if err := n.validateAdjudicatorContract(adjAddr); err == nil {
+					n.adjAddr = adjAddr
+					n.adjudicatorSetup = true
+					fmt.Printf("✅ Adjudicator contract validated at %x\n", adjAddr)
+					return nil
+				}
+			}
+
+			fmt.Printf("❌ Adjudicator contract validation for %v failed\n", token)
```

Comment:
Please always print the error messages if we have them, otherwise they are wasted.

Reply:
not applicable to most recent version

Reply:
in the most recent version, the errors are returned to the calling function. is that sufficient? if not, could you give a concrete example how you would like the error reporting to be done?

Reply:
Yes this is fine, just wanted to say that we should always use the error messages that we have.